File: backend/src/controllers/clienteController.js
const mongoose = require('mongoose');
const Cliente = require('../models/Cliente');
const User = require('../../models/User');
const { registrarLog } = require('./logController');

const ADMIN_EMAIL_FALLBACK = 'diegorafa87@gmail.com';

function resolveRole(user) {
  if (!user) return 'ADMIN';
  if (user?.role) return user.role;
  return user?.email === ADMIN_EMAIL_FALLBACK ? 'ADMIN' : 'FILHO';
}

async function getUserFromHeader(req) {
  const email = req.headers['x-user-email'];
  if (!email) return null;
  return User.findOne({ email });
}

function userCanAccessCliente(user, cliente) {
  const role = resolveRole(user);
  if (role === 'ADMIN') return true;
  if (role === 'FILHO') return (cliente.consultoria || '') === (user?.consultoria || '');
  if (role === 'NETO') return String(user?.clienteId || '') === String(cliente?._id || '');
  return false;
}

exports.cadastrarCliente = async (req, res) => {
  try {
    let { razaoSocial, cnpj, email, telefone, consultoria } = req.body;
    const user = await getUserFromHeader(req);
    const role = resolveRole(user);

    if (role === 'NETO') {
      return res.status(403).json({ error: 'Usuário NETO não pode cadastrar clientes.' });
    }

    if (role === 'FILHO') {
      consultoria = user.consultoria;
    }

    if (!razaoSocial || !cnpj || !email || !telefone || !consultoria) {
      return res.status(400).json({ error: 'Preencha todos os campos!' });
    }

    const existeCnpj = await Cliente.findOne({ cnpj });
    if (existeCnpj) return res.status(409).json({ error: 'CNPJ já cadastrado!' });

    const existeRazao = await Cliente.findOne({ razaoSocial: { $regex: `^${razaoSocial}$`, $options: 'i' } });
    if (existeRazao) return res.status(409).json({ error: 'Razão Social já cadastrada!' });

    const novoCliente = await Cliente.create({ razaoSocial, cnpj, email, telefone, consultoria });
    registrarLog('CADASTRAR_CLIENTE', cnpj, { razaoSocial, email, telefone, consultoria });
    res.status(201).json(novoCliente);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao cadastrar cliente', details: err.message });
  }
};

exports.listarClientes = async (req, res) => {
  try {
    const { consultoria } = req.query;
    let filtro = {};

    const user = await getUserFromHeader(req);
    const role = resolveRole(user);

    if (role === 'FILHO') {
      filtro.consultoria = user.consultoria;
    }

    if (role === 'NETO') {
      if (!user?.clienteId) return res.json([]);
      filtro._id = user.clienteId;
    }

    if (consultoria) {
      filtro.consultoria = consultoria;
    }

    const clientes = await Cliente.find(filtro);
    res.json(clientes);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao listar clientes', details: err.message });
  }
};

exports.atualizarStatus = async (req, res) => {
  try {
    const { id } = req.params;
    const { status } = req.body;

    if (!mongoose.Types.ObjectId.isValid(id)) {
      return res.status(400).json({ error: 'ID de cliente inválido' });
    }

    if (!['NOVO', 'ATIVO', 'CORRIGIR', 'SUSPENSO'].includes(status)) {
      return res.status(400).json({ error: 'Status inválido' });
    }

    if (status === 'NOVO') {
      return res.status(400).json({ error: 'Status NOVO só pode ser atribuído no cadastro do cliente.' });
    }

    const user = await getUserFromHeader(req);
    const clienteAtual = await Cliente.findById(id);
    if (!clienteAtual) return res.status(404).json({ error: 'Cliente não encontrado' });

    if (!userCanAccessCliente(user, clienteAtual)) {
      return res.status(403).json({ error: 'Acesso negado para alterar status deste cliente.' });
    }

    const cliente = await Cliente.findByIdAndUpdate(id, { status }, { new: true });
    registrarLog('ATUALIZAR_STATUS', cliente.cnpj, { id, status });
    res.json(cliente);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao atualizar status', details: err.message });
  }
};

exports.detalharCliente = async (req, res) => {
  try {
    const { id } = req.params;

    if (!mongoose.Types.ObjectId.isValid(id)) {
      return res.status(400).json({ error: 'ID de cliente inválido' });
    }

    const user = await getUserFromHeader(req);
    const cliente = await Cliente.findById(id);
    if (!cliente) return res.status(404).json({ error: 'Cliente não encontrado' });

    if (!userCanAccessCliente(user, cliente)) {
      return res.status(403).json({ error: 'Acesso negado para este cliente.' });
    }

    res.json(cliente);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao detalhar cliente', details: err.message });
  }
};

exports.excluirCliente = async (req, res) => {
  try {
    const { id } = req.params;

    if (!mongoose.Types.ObjectId.isValid(id)) {
      return res.status(400).json({ error: 'ID de cliente inválido' });
    }

    const user = await getUserFromHeader(req);
    const clienteAtual = await Cliente.findById(id);
    if (!clienteAtual) return res.status(404).json({ error: 'Cliente não encontrado' });

    if (!userCanAccessCliente(user, clienteAtual)) {
      return res.status(403).json({ error: 'Acesso negado para excluir este cliente.' });
    }

    await Cliente.findByIdAndDelete(id);
    registrarLog('EXCLUIR_CLIENTE', id, {});
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ error: 'Erro ao excluir cliente', details: err.message });
  }
};

exports.editarCliente = async (req, res) => {
  try {
    const { id } = req.params;
    let { razaoSocial, cnpj, email, telefone, consultoria } = req.body;

    if (!mongoose.Types.ObjectId.isValid(id)) {
      return res.status(400).json({ error: 'ID de cliente inválido' });
    }

    const user = await getUserFromHeader(req);
    const role = resolveRole(user);

    const clienteAtual = await Cliente.findById(id);
    if (!clienteAtual) return res.status(404).json({ error: 'Cliente não encontrado' });

    if (!userCanAccessCliente(user, clienteAtual)) {
      return res.status(403).json({ error: 'Acesso negado para editar este cliente.' });
    }

    if (role === 'FILHO') {
      consultoria = user.consultoria;
    }

    if (!razaoSocial || !cnpj || !email || !telefone || !consultoria) {
      return res.status(400).json({ error: 'Preencha todos os campos!' });
    }

    const cliente = await Cliente.findByIdAndUpdate(
      id,
      { razaoSocial, cnpj, email, telefone, consultoria },
      { new: true }
    );

    registrarLog('EDITAR_CLIENTE', cnpj, { id, razaoSocial, email, telefone, consultoria });
    res.json(cliente);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao editar cliente', details: err.message });
  }
};

exports.atualizarObservacao = async (req, res) => {
  try {
    const { id } = req.params;
    const { observacao } = req.body;

    if (!mongoose.Types.ObjectId.isValid(id)) {
      return res.status(400).json({ error: 'ID de cliente inválido' });
    }

    const user = await getUserFromHeader(req);
    const clienteAtual = await Cliente.findById(id);
    if (!clienteAtual) return res.status(404).json({ error: 'Cliente não encontrado' });

    if (!userCanAccessCliente(user, clienteAtual)) {
      return res.status(403).json({ error: 'Acesso negado para atualizar observação deste cliente.' });
    }

    const cliente = await Cliente.findByIdAndUpdate(id, { observacao }, { new: true });
    registrarLog && registrarLog('ATUALIZAR_OBSERVACAO', cliente.cnpj, { id, observacao });
    res.json(cliente);
  } catch (err) {
    res.status(500).json({ error: 'Erro ao atualizar observação', details: err.message });
  }
};

File: python_backend/controllers/cliente_controller.py
from fastapi import HTTPException
from bson import ObjectId
from pymongo import ReturnDocument
from db import db
from controllers.log_controller import registrar_log

VALID_STATUSES = ['NOVO','ATIVO','CORRIGIR','SUSPENSO']

async def cadastrar_cliente(payload: dict):
    razaoSocial = payload.get('razaoSocial')
    cnpj = payload.get('cnpj')
    email = payload.get('email')
    telefone = payload.get('telefone')
    consultoria = payload.get('consultoria')
    if not all([razaoSocial, cnpj, email, telefone, consultoria]):
        raise HTTPException(status_code=400, detail='Preencha todos os campos!')
    coll = db['clientes']
    existe_cnpj = await coll.find_one({'cnpj': cnpj})
    if existe_cnpj:
        raise HTTPException(status_code=409, detail='CNPJ já cadastrado!')
    existe_razao = await coll.find_one({'razaoSocial': {'$regex': f'^{razaoSocial}$', '$options': 'i'}})
    if existe_razao:
        raise HTTPException(status_code=409, detail='Razão Social já cadastrada!')
    novo = { 'razaoSocial': razaoSocial, 'cnpj': cnpj, 'email': email, 'telefone': telefone, 'consultoria': consultoria }
    res = await coll.insert_one(novo)
    novo['_id'] = str(res.inserted_id)
    registrar_log('CADASTRAR_CLIENTE', cnpj, {'razaoSocial': razaoSocial, 'email': email, 'telefone': telefone, 'consultoria': consultoria})
    return novo

async def listar_clientes(consultoria: str = None):
    filtro = {}
    if consultoria:
        filtro['consultoria'] = consultoria
    coll = db['clientes']
    try:
        cursor = coll.find(filtro)
        out = []
        async for doc in cursor:
            doc['_id'] = str(doc.get('_id'))
            out.append(doc)
        return out
    except Exception:
        # Se o Mongo estiver indisponível, retornar lista vazia para permitir funcionamento do frontend em dev
        return []


def _validate_objectid(id_str: str):
    try:
        return ObjectId(id_str)
    except Exception:
        raise HTTPException(status_code=400, detail='ID de cliente inválido')

async def atualizar_status(id: str, payload: dict):
    status = payload.get('status')
    if status not in VALID_STATUSES:
        raise HTTPException(status_code=400, detail='Status inválido')
    if status == 'NOVO':
        raise HTTPException(status_code=400, detail='Status NOVO só pode ser atribuído no cadastro do cliente.')
    coll = db['clientes']
    oid = _validate_objectid(id)
    res = await coll.find_one_and_update({'_id': oid}, {'$set': {'status': status}}, return_document=ReturnDocument.AFTER)
    if not res:
        raise HTTPException(status_code=404, detail='Cliente não encontrado')
    registrar_log('ATUALIZAR_STATUS', res.get('cnpj'), {'id': id, 'status': status})
    res['_id'] = str(res.get('_id'))
    return res

async def detalhar_cliente(id: str):
    oid = _validate_objectid(id)
    coll = db['clientes']
    doc = await coll.find_one({'_id': oid})
    if not doc:
        raise HTTPException(status_code=404, detail='Cliente não encontrado')
    doc['_id'] = str(doc.get('_id'))
    return doc

async def excluir_cliente(id: str):
    oid = _validate_objectid(id)
    coll = db['clientes']
    res = await coll.find_one_and_delete({'_id': oid})
    if not res:
        raise HTTPException(status_code=404, detail='Cliente não encontrado')
    registrar_log('EXCLUIR_CLIENTE', id, {})
    return {'success': True}

async def editar_cliente(id: str, payload: dict):
    razaoSocial = payload.get('razaoSocial')
    cnpj = payload.get('cnpj')
    email = payload.get('email')
    telefone = payload.get('telefone')
    consultoria = payload.get('consultoria')
    if not all([razaoSocial, cnpj, email, telefone, consultoria]):
        raise HTTPException(status_code=400, detail='Preencha todos os campos!')
    oid = _validate_objectid(id)
    coll = db['clientes']
    res = await coll.find_one_and_update({'_id': oid}, {'$set': {'razaoSocial': razaoSocial, 'cnpj': cnpj, 'email': email, 'telefone': telefone, 'consultoria': consultoria}}, return_document=ReturnDocument.AFTER)
    if not res:
        raise HTTPException(status_code=404, detail='Cliente não encontrado')
    registrar_log('EDITAR_CLIENTE', cnpj, {'id': id, 'razaoSocial': razaoSocial, 'email': email, 'telefone': telefone, 'consultoria': consultoria})
    res['_id'] = str(res.get('_id'))
    return res

async def atualizar_observacao(id: str, payload: dict):
    observacao = payload.get('observacao')
    if observacao is None:
        raise HTTPException(status_code=400, detail='Observação é obrigatória')
    oid = _validate_objectid(id)
    coll = db['clientes']
    res = await coll.find_one_and_update({'_id': oid}, {'$set': {'observacao': observacao}}, return_document=ReturnDocument.AFTER)
    if not res:
        raise HTTPException(status_code=404, detail='Cliente não encontrado')
    registrar_log('ATUALIZAR_OBSERVACAO', res.get('cnpj'), {'id': id, 'observacao': observacao})
    res['_id'] = str(res.get('_id'))
    return res
